### PR TITLE
TN-2627 encounter api & allow pre-registered users in via url authentication

### PR DESCRIPTION
**Pull request by pbugni**
NB - to enable, set config value `ENABLE_URL_AUTHENTICATED` to a non false value.

- TN-2636: Add endpoint for front-end to acquire logged in user's current encounter.  The `auth_method` will define features that should be hidden from the user, such as making profile changes when `auth_method == 'url_authenticated'`
 
- TN-2635: Entry via /access/<token> now allows EPROMs users direct access, providing they've registered and finished initial queries.  The encounter will have the aforementioned `auth_method == 'url_authenticated'`

- TN-2647: Add css-class "portal-weak-auth-disabled" to the completed assessment tile, so the front end can disable or redirect to login, when the user's current encounter has `auth_method == 'url_authenticated'`

- TN-2648: Consider `url_authenticated` when evaluating authorization for all API endpoints

### CODE REVIEW

File: portal/views/user.py [hunk 1]
```diff
@@ -957,6 +957,89 @@ def delete_user_consents(user_id):
     return jsonify(message="ok")
 
 
+@user_api.route('/user/<int:user_id>/encounter', methods=('GET',))
```
Comment (achen2401) [inline portal/views/user.py:960]: @pbugni Paul, is the `user_id` required if only current user is expected?  I think that the frontend will need to get the user id (e.g. maybe call to /api/me) before it calls this api endpoint, just thinking that it probably doesn't need to make that extra call to get the user id.

File: portal/templates/portal_wrapper.html [hunk 2]
```diff
@@ -94,7 +94,7 @@
                          <button id="tnthUserBtn" type="button" class="tnth-nav-btn tnth-white tnth-dropdown-toggle" data-toggle="tnth-dropdown">
                             {% if enable_links %}&nbsp;&nbsp; {{ _("MENU") }} <i class="fa fa-bars"></i>{% endif %}
                             <img class="profile-img" src="{{movember_profile}}" width="50" height="50" alt="{{_('Profile image')}}"/>
-                            <span class="welcome-text"><em>{{ _("Welcome") }}</em>{% if user.display_name != "Anonymous" %}, {{ user.display_name }}{% endif %}</span>
+                            <span class="welcome-text portal-weak-auth-hide"><em>{{ _("Welcome") }}</em>{% if user.display_name != "Anonymous" %}, {{ user.display_name }}{% endif %}</span>
```
Comment (ivan-c) [inline portal/templates/portal_wrapper.html:97]: Would a check for `user.has_role('write_only')` be more appropriate than checking a symptom of that role?

File: portal/views/assessment_engine.py [hunk 3]
```diff
@@ -1623,6 +1618,11 @@ def present_needed():
         return redirect('/')
 
     url = url_for('.present_assessment', **args)
+
+    encounter = current_user().current_encounter
+    if encounter.auth_method == "url_authenticated":
+        return redirect('/confirm-identity?redirect_url={}'.format(url))
```
Comment (ivan-c) [inline portal/views/assessment_engine.py:1624]: I've found redirects really hard to debug, would it make sense to add a logging line here before redirecting?

File: portal/models/user.py [hunk 4]
```diff
@@ -1643,8 +1647,19 @@ def check_role(self, permission, other_id):
...
 
+        NB - a user with "url_authenticated" as their current encounter's
+        auth_type will NOT have any access, unless specifically requested
+        via the "allow_on_url_authenticated_encounters" parameter
+
         """
         assert (permission in ('view', 'edit'))  # limit vocab for now
+        if (
+                not allow_on_url_authenticated_encounters and
+                current_app.config.get('ENABLE_URL_AUTHENTICATED') and
```
Comment (ivan-c) [inline portal/models/user.py:1658]: Is there a `os.environ.get('ENABLE_URL_AUTHENTICATED')` call in [config.py](https://github.com/uwcirg/truenth-portal/blob/develop/portal/config/config.py) for this? (if we want to config via docker)

File: portal/views/assessment_engine.py (hunk 3 above)
Comment (achen2401) [reply portal/views/assessment_engine.py:1625]: thanks, ivan, added

File: portal/templates/portal_wrapper.html (hunk 2 above)
Comment (achen2401) [reply portal/templates/portal_wrapper.html:97]: think the issue here that we don't check the user's role, but checking the authentication method, i.e. url_authenticated, in the user's encounter instead

File: portal/models/user.py (hunk 4 above)
Comment (pbugni) [reply portal/models/user.py:1658]: i expect this to become part of eproms' site_persistence config, but given they might want to toggle before a release cycle - yes, we'll need that!

File: portal/templates/portal_wrapper.html (hunk 2 above)
Comment (pbugni) [reply portal/templates/portal_wrapper.html:97]: @achen2401 is correct.  roles are generally static, as opposed to an encounter level auth method.

File: portal/models/user.py (hunk 4 above)
Comment (ivan-c) [reply portal/models/user.py:1658]: Added in 90156217c91cb1b151bc6834f99ff6f2b22256fc

File: portal/templates/portal_wrapper.html (hunk 2 above)
Comment (ivan-c) [reply portal/templates/portal_wrapper.html:97]: Ah, will the check in the view also work in the template?:
`user.current_encounter.auth_method == 'url_authenticated'`

Comment (achen2401) [reply portal/templates/portal_wrapper.html]: thanks, ivan, I think it works, although, in this case we want to hide the relevant display `user.current_encounter.auth_method != 'url_authenticated'`

**Pull request merged**